src/main/java/xyz/nkomarn/harbor/listener/AfkListener.java

```diff
@@ -0,0 +1,153 @@
+    }
+
+    /**
+     * Provides a way to start the listener
+     */
+    public void start() {
+        JavaPlugin plugin = JavaPlugin.getPlugin(Harbor.class);
+        playerQueue = new ArrayDeque<>();
+        movementChecker = new PlayerMovementChecker();
+
+        playerQueue.addAll(Bukkit.getOnlinePlayers().stream().map((Function<Player, AfkPlayer>) AfkPlayer::new).collect(Collectors.toSet()));
+
+        // Register listeners after populating the queue
+        Bukkit.getServer().getPluginManager().registerEvents(this, plugin);
+
+        // We want every player to get a check every 20 ticks.
+        movementChecker.runTaskTimer(plugin, 20, 20);
+
+        JavaPlugin.getPlugin(Harbor.class).getLogger().info("Fallback AFK detection system is enabled");
+    }
+
+    /**
+     * Provides a way to halt the listener
+     */
+    public void stop() {
+        movementChecker.cancel();
+        HandlerList.unregisterAll(this);
+        playerQueue = null;
+        JavaPlugin.getPlugin(Harbor.class).getLogger().info("Fallback AFK detection system is disabled");
+    }
+
+    @EventHandler(ignoreCancelled = true)
+    public void onChat(AsyncPlayerChatEvent event) {
+        afkProvider.updateActivity(event.getPlayer());
+    }
+
+    @EventHandler(ignoreCancelled = true)
+    public void onCommand(PlayerCommandPreprocessEvent event) {
+        afkProvider.updateActivity(event.getPlayer());
+    }
+
+    @EventHandler(ignoreCancelled = true)
+    public void onInventoryClick(InventoryClickEvent event) {
+        afkProvider.updateActivity((Player) event.getWhoClicked());
+    }
+
+    @EventHandler
+    public void onJoin(PlayerJoinEvent event) {
+        playerQueue.add(new AfkPlayer(event.getPlayer()));
+        afkProvider.updateActivity(event.getPlayer());
+    }
+
+    @EventHandler
+    public void onLeave(PlayerQuitEvent event) {
+        playerQueue.remove(new AfkPlayer(event.getPlayer()));
+        afkProvider.removePlayer(event.getPlayer().getUniqueId());
+    }
+
+    /**
+     * Internal class for handling the task of checking player movement; Is a separate task so that we can cancel and restart it easily
+     */
+    private class PlayerMovementChecker extends BukkitRunnable {
+        @Override
+        public synchronized void run() {
+            // Short circuit exit if there aren't any players to check
+            if (playerQueue.isEmpty())
+                return;
+
+
+            AfkPlayer firstPlayer = playerQueue.peek();
+            AfkPlayer player = firstPlayer;
+
+            long start = System.currentTimeMillis();
+            do {
+                if (player.changed()) {
+                    afkProvider.updateActivity(player.player);
+                }
+                playerQueue.add(player);
+            } while (System.currentTimeMillis() - start < MAX_PROCESS_TIME_MS &&
```

Comment:
Just noticed this.

This is exactly what I wanted to avoid. Iterating over all players each tick. Also the synchronized doesnt makes any sense at all since this task is executed synced by the mainthread. There is no way that this will ever be executed twice.
This is all just very over complicated without any advance in functionallity. Its even worth then before.
My mechanic checked every player every 20 ticks, which is by far enough. The new mechanic checks every player every thick, which is way to much and is not required at all.
The do while makes it also overly complex.
Also I would remove the max process time mechanic since it is not really required like stated in #87.

Reply:
Also revisiting you code this is a real mess.

You peek a player. Then you put the player back to the queue which results in the player being twice in the queue.
After this you poll a player without knowing that the player will be added back into the queue, which will result in a loss of players in the queue.

Please just change this back to my routine. You can also remove the 20 ms mechanic since this condition will never be hit unless over 1000 players are on a server.

Also remove the useless synchronize keywords, which makes no sense at all in a blocking task in the main thread. You meant it good, but made it worse sadly. Sometime simple is better.

Reply:
I miswrote it while far too tired and should not have pushed, I;m sorry.